File: polyglot/envelope_v1/typescript/main.ts
// B1 fourteen-language conformance: TypeScript.
//
// Invariant owned: schema_version_is_exact
//
// A reader that accepts an unknown schema_version is processing a record whose
// meaning it has not established. B1's envelope readers refuse rather than
// guess, and this consumer checks the committed vector carries the exact
// version those readers are written against. Exact equality, not a prefix or a
// range: "b1-event-envelope-10" starts with "b1-event-envelope-1".
//
// TypeScript owns it because TypeScript is B1's contract language for the
// desktop surface, where an envelope arrives from a backend that may be a
// different build than the UI.
//
// digest_verification: STDLIB. The Node runtime provides node:crypto.
//
// The ambient declarations below are deliberate: this consumer must compile
// with a bare `tsc` and no @types/node, because the harness requires each
// language to build from a single source file with nothing installed beyond its
// own toolchain.
//
// Usage: node main.js <contract.json> <vector.canonical> <expected.json>

// `console` is deliberately not declared here: tsc's default lib already
// supplies it, and redeclaring it is an error rather than a convenience.
declare function require(name: string): any;
declare const process: { argv: string[]; exit(code: number): never };

const { createHash } = require("node:crypto");
const { readFileSync } = require("node:fs");

const INVARIANT = "schema_version_is_exact";
const EXPECTED_SCHEMA = "b1-event-envelope-1";

interface Contract {
  invariants?: Record<string, boolean>;
}

interface Expected {
  canonical_sha256?: string;
}

function main(argv: string[]): number {
  if (argv.length !== 3) {
    console.error("usage: node main.js <contract.json> <vector.canonical> <expected.json>");
    return 2;
  }
  const contractPath = argv[0];
  const vectorPath = argv[1];
  const expectedPath = argv[2];

  const contract: Contract = JSON.parse(readFileSync(contractPath, "utf8"));
  const vectorBytes = readFileSync(vectorPath);
  const expected: Expected = JSON.parse(readFileSync(expectedPath, "utf8"));

  // The contract must actually assert this invariant. A consumer checking a
  // property nobody declared is testing its own opinion, not B1's contract.
  if (contract.invariants?.[INVARIANT] !== true) {
    console.error(`TypeScript: contract does not assert ${INVARIANT}`);
    return 4;
  }

  const envelope: Record<string, unknown> = JSON.parse(vectorBytes.toString("utf8"));
  const version = envelope["schema_version"];

  if (typeof version !== "string") {
    console.error("TypeScript: schema_version is absent or not a string");
    return 5;
  }
  if (version !== EXPECTED_SCHEMA) {
    console.error(
      `TypeScript: schema_version is ${JSON.stringify(version)}, expected ` +
        `${JSON.stringify(EXPECTED_SCHEMA)}. A reader that accepts an unknown version is ` +
        `processing a record whose meaning it has not established.`,
    );
    return 5;
  }

  const actual = createHash("sha256").update(vectorBytes).digest("hex");
  if (actual !== expected.canonical_sha256) {
    console.error(
      `TypeScript: digest mismatch; declared ${expected.canonical_sha256} ` +
        `but the bytes hash to ${actual}`,
    );
    return 5;
  }

  console.log(`TYPESCRIPT:POSTCONDITION:${INVARIANT}`);
  return 0;
}

process.exit(main(process.argv.slice(2)));

File: polyglot/envelope_v1/kotlin/Main.kt
// B1 fourteen-language conformance: Kotlin.
//
// Invariant owned: epoch_is_non_negative_integer
//
// The epoch is B1's fencing token. Every stale-attempt rejection in the root
// journal is a comparison against it, so it has to be an integer that compares
// monotonically. A fractional epoch cannot be compared reliably across
// languages for the same float-rendering reason B1 bans floats outright, and a
// negative epoch has no meaning: fences start at zero and only ever move up.
//
// Kotlin owns it because Kotlin is B1's Android surface, and a mobile client
// that misreads a fence would replay a superseded attempt against live state.
//
// digest_verification: STDLIB. java.security.MessageDigest is available on the
// JVM, so this consumer also verifies the declared digest.
//
// Usage: java -jar consumer.jar <contract.json> <vector.canonical> <expected.json>

import java.io.File
import java.security.MessageDigest
import kotlin.system.exitProcess

private const val INVARIANT = "epoch_is_non_negative_integer"

private fun failWith(message: String, code: Int): Nothing {
    System.err.println("Kotlin: $message")
    exitProcess(code)
}

/**
 * Pull a top-level scalar out of canonical JSON without a parser.
 *
 * Canonical form has no insignificant whitespace, so a top-level key is always
 * followed immediately by ':' and then the value. Scanning is used rather than
 * adding a JSON dependency: this consumer must build from a single source file
 * with nothing but the standard library.
 */
private fun scalarAfterKey(text: String, key: String): String? {
    val needle = "\"$key\":"
    val at = text.indexOf(needle)
    if (at < 0) return null
    val start = at + needle.length
    var end = start
    while (end < text.length && text[end] !in ",}]") end++
    return text.substring(start, end).trim()
}

fun main(args: Array<String>) {
    if (args.size != 3) {
        failWith("usage: consumer <contract.json> <vector.canonical> <expected.json>", 2)
    }

    val contract = File(args[0]).readText(Charsets.UTF_8)
    val vectorBytes = File(args[1]).readBytes()
    val expected = File(args[2]).readText(Charsets.UTF_8)
    val vector = String(vectorBytes, Charsets.UTF_8)

    // The contract must actually assert this invariant. A consumer checking a
    // property nobody declared is testing its own opinion, not B1's contract.
    if (!contract.contains("\"$INVARIANT\": true") && !contract.contains("\"$INVARIANT\":true")) {
        failWith("contract does not assert $INVARIANT", 4)
    }

    val raw = scalarAfterKey(vector, "epoch")
        ?: failWith("vector carries no epoch; every envelope must carry a fence", 5)

    if (raw.contains('.') || raw.contains('e') || raw.contains('E')) {
        failWith(
            "epoch $raw is not an integer literal. A fractional fence cannot be compared " +
                "reliably across languages, for the same reason B1 bans floats outright.",
            5
        )
    }
    val epoch = raw.toLongOrNull()
        ?: failWith("epoch $raw is not a valid integer", 5)
    if (epoch < 0) {
        failWith("epoch $epoch is negative; fences start at zero and only ever move up", 5)
    }

    val declared = Regex("\"canonical_sha256\"\\s*:\\s*\"([0-9a-f]{64})\"")
        .find(expected)?.groupValues?.get(1)
        ?: failWith("expected.json declares no canonical_sha256", 5)

    val actual = MessageDigest.getInstance("SHA-256")
        .digest(vectorBytes)
        .joinToString("") { "%02x".format(it) }
    if (actual != declared) {
        failWith("digest mismatch; declared $declared but the bytes hash to $actual", 5)
    }

    println("KOTLIN:POSTCONDITION:$INVARIANT")
}
